scene/3d/portal.cpp

```diff
@@ -0,0 +1,642 @@
+			VisualServer::get_singleton()->portal_set_scenario(_portal_rid, get_world()->get_scenario());
+
+			// we can't calculate world points until we have entered the tree
+			portal_update();
+			update_gizmo();
+
+		} break;
+		case NOTIFICATION_EXIT_WORLD: {
+			// partially destroy  the visual server portal when the editor portal exits the scene tree
+			VisualServer::get_singleton()->portal_set_scenario(_portal_rid, RID());
+		} break;
+		case NOTIFICATION_TRANSFORM_CHANGED: {
+			// keep the world points and the visual server up to date
+			portal_update();
+		} break;
+	}
+}
+
+void Portal::set_portal_active(bool p_active) {
+	_settings_active = p_active;
+	VisualServer::get_singleton()->portal_set_active(_portal_rid, p_active);
+}
+
+bool Portal::get_portal_active() const {
+	return _settings_active;
+}
+
+void Portal::set_use_default_margin(bool p_use) {
+	_use_default_margin = p_use;
+	update_gizmo();
+}
+
+bool Portal::get_use_default_margin() const {
+	return _use_default_margin;
+}
+
+void Portal::set_portal_margin(real_t p_margin) {
+	_margin = p_margin;
+
+	if (!_use_default_margin) {
+		// give visual feedback in the editor for the portal margin zone
+		update_gizmo();
+	}
+}
+
+real_t Portal::get_portal_margin() const {
+	return _margin;
+}
+
+void Portal::resolve_links(const RID &p_from_room_rid) {
+	Room *linkedroom = nullptr;
+	if (has_node(_settings_path_linkedroom)) {
+		linkedroom = Object::cast_to<Room>(get_node(_settings_path_linkedroom));
+	}
+
+	if (linkedroom) {
+		_linkedroom_ID[1] = linkedroom->_room_ID;
+
+		// send to visual server
+		VisualServer::get_singleton()->portal_link(_portal_rid, p_from_room_rid, linkedroom->_room_rid, _settings_two_way);
+	} else {
+		_linkedroom_ID[1] = -1;
+	}
+}
+
+void Portal::set_linked_room_internal(const NodePath &link_path) {
+	_settings_path_linkedroom = link_path;
+}
+
+bool Portal::try_set_unique_name(const String &p_name) {
+	SceneTree *scene_tree = get_tree();
+	if (!scene_tree) {
+		// should not happen in the editor
+		return false;
+	}
+
+	Viewport *root = scene_tree->get_root();
+	if (!root) {
+		return false;
+	}
+
+	Node *found = root->find_node(p_name, true, false);
+
+	// if the name does not already exist in the scene tree, we can use it
+	if (!found) {
+		set_name(p_name);
+		return true;
+	}
+
+	// we are trying to set the same name this node already has...
+	if (found == this) {
+		// noop
+		return true;
+	}
+
+	return false;
+}
+
+void Portal::set_linked_room(const NodePath &link_path) {
+	// change the name of the portal as well, if the link looks legit
+	Room *linkedroom = nullptr;
+	if (has_node(link_path)) {
+		linkedroom = Object::cast_to<Room>(get_node(link_path));
+
+		if (linkedroom) {
+			if (linkedroom != get_parent()) {
+				_settings_path_linkedroom = link_path;
+
+				// change the portal name
+				String szLinkRoom = RoomManager::_find_name_after(linkedroom, "Room");
```

Comment:
What does `sz` stand for? Should probably use snake_case for locals in this method, unless they're referring to names in a thirdparty API.

Reply:
Ah this is a leftover from LPortal which didn't use snake_case. I'll change.

`sz` I shouldn't be using either (it is hungarian notation for null terminated string), so tell me off when you see it. :smile:

Reply:
I've done a find_in_files and replaced all the `sz` I used in the PR. :+1: